### PR TITLE
DTaaS CLI: Base structure and User add/Delete for Compose 

### PR DESCRIPTION
**Pull request by astitva1905**
- Creates new file for additional users ml-workspaces 
- Demos additions of a new user, across all relevant files

Tests:
Screenshot from 2024-03-03 13-13-52
Screenshot from 2024-03-03 13-13-33
Screenshot from 2024-03-03 13-13-28
Screenshot from 2024-03-03 13-13-21

To do:
- Adding rules to conf file in user add -> (for next PR) 
- Starting new user container from within the user add cli (docker compose -f compose.users.yml up <username> )  -> Done
- No hardcoding config, instead users.yml.yemplate with recursive replace to fill in params -> Done 
- File path decision for global config file (~/.config/dtaas/config.yml) -> Changed to dtaas.toml (fornow assumed as always exists in DTaaS/cli/); Done.

- Setting networks across compose files 
- Linting and testing
- Packaging and publishing cli

### CODE REVIEW

File: docker/.env [hunk 1]
```diff
@@ -6,4 +6,5 @@ CLIENT_SECRET='xx'
 OAUTH_SECRET='random-secret-string'
 CLIENT_CONFIG='/home/Desktop/DTaaS/deploy/config/client/env.local.js'
 username1='user1'
-username2='user2'
\ No newline at end of file
+username2='user2'
```
Comment (prasadtalasila) [inline docker/.env:9]: please keep two usernames as default.

**Comment by prasadtalasila** [comment]
@astitva1905 
Please use [click](https://github.com/pallets/click) for CLI development.

**Comment by prasadtalasila** [comment]
@astitva1905 
There is in-built support for TOML config format in python 3.11 onwards. Let us go with TOML config format (instead of YAML format) for DTaaS CLI. The suggested config is in [wiki](https://github.com/INTO-CPS-Association/DTaaS/wiki/Config). Only the following part needs implementation in this PR.

```toml
# This is the config for DTaaS CLI

name = "Digital Twin as a Service (DTaaS)"
version = "0.1"
owner = "The INTO-CPS-Association"
git-repo = "https://github.com/into-cps-association/DTaaS.git"

[common]
# absolute path to the DTaaS application directory
path = "/home/dtaas/app"
```

Please name this file as `dtaas.toml`. We can postpone the default location of this file from this PR and always assume that `dtaas.toml` exists in the execution directory of `dtaas` CLI. If the file does not exist the CLI can create the same.

The format of the config subcommands are:

```sh
# inspiration "git config -h"
dtaas config --list
dtaas config --get path
dtaas config --set path "<path>"
dtaas config --unset path "<path>"
```

File: cli/dtaas_cli/cmd.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+@admin.group()
+def user():
+    """user management commands"""
+    pass
+
+#### admin group commands
+@admin.command()
+@click.argument('path', type=click.Path(exists=True))
+def set_path(path):
```
Comment (prasadtalasila) [inline cli/dtaas_cli/cmd.py:25]: please use camel case.

File: cli/dtaas_cli/pkg/dtaas.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from . import helper
```
Comment (prasadtalasila) [inline cli/dtaas_cli/pkg/dtaas.py:1]: perhaps this file can be renamed as `config.py`?

File: cli/dtaas_cli/pkg/dtaas.py
Comment (prasadtalasila) [inline cli/dtaas_cli/pkg/dtaas.py:1]: please rename `dtaas_cli` directory to `src`.

File: cli/dtaas_cli/pkg/helper.py
Comment (prasadtalasila) [inline cli/dtaas_cli/pkg/helper.py:1]: convention is to call this file `util.py`

File: cli/dtaas_cli/pkg/helper.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import yaml
```
Comment (prasadtalasila) [inline cli/dtaas_cli/pkg/helper.py:1]: This file can take the responsibility for parsing the `dtaas.toml` file.

File: cli/setup.py
Comment (prasadtalasila) [inline cli/setup.py:1]: please use the project structure recommended by the poetry package manager.

File: docker/.env [hunk 5]
```diff
@@ -6,4 +6,5 @@ CLIENT_SECRET='xx'
 OAUTH_SECRET='random-secret-string'
```
Comment (prasadtalasila) [inline docker/.env:6]: This file can be auto-generated and managed by `dtaas config` subcommand. The source of truth is the `dtaas.toml`.

File: docker/conf.local
Comment (prasadtalasila) [inline docker/conf.local:1]: The traefik forward-auth config files can be auto-generated and managed by `dtaas config` subcommand. The source of truth is the `dtaas.toml`.

File: cli/dtaas_cli/pkg/dtaas.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
```
Comment (prasadtalasila) [inline cli/dtaas_cli/pkg/dtaas.py:2]: Please add a **Config** class and have its constructor receive the `dtaas.toml` config file.

File: docker/compose.users.yml
Comment (prasadtalasila) [inline docker/compose.users.yml:1]:
```yaml
version: '3'
services:
  user:
    image: mltooling/ml-workspace-minimal:0.13.2
    volumes:
      - ${DTAAS_DIR}/files/common:/workspace/common
      - ${DTAAS_DIR}/files/${username}:/workspace
    environment:
      - AUTHENTICATE_VIA_JUPYTER=
      - WORKSPACE_BASE_URL=${username}
    shm_size: 512m
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.u3.entryPoints=web"
      - "traefik.http.routers.u3.rule=PathPrefix(`/${username}`)"
      - "traefik.http.routers.u3.middlewares=traefik-forward-auth"
```

**Comment by prasadtalasila** [comment]
The additional stack to be put in:

* poetry project structure
* pytest for tests
* pylint for linting
* pyright for type checking

**Additional TODO**: github workflow file to perform all the poetry steps including `publish --dry-run` step.

**Comment by prasadtalasila** [comment]
Let the forward-auth config update step (for conf.server and conf.local) be manual.
Use https://pypi.org/project/tomlkit/ for toml file parsing.

File: cli/dtaas.toml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# This is the config for DTaaS CLI
+
+name = "Digital Twin as a Service (DTaaS)"
+version = "0.1"
+owner = "The INTO-CPS-Association"
+git-repo = "https://github.com/into-cps-association/DTaaS.git"
+
+[common]
+# absolute path to the DTaaS application directory
+server-dns = "domain.com"
```
Comment (prasadtalasila) [inline cli/dtaas.toml:10]: foo.com

File: cli/dtaas.toml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+path = "/home/astitva/Desktop/yamlstuff"
```
Comment (prasadtalasila) [inline cli/dtaas.toml:11]: /home/dtaas/DTaaS

File: cli/dtaas.toml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+[users]
+# matching user info must present in this config file
+add = ["astitvasehgal05","astitvasehgal19", "prasad"]
```
Comment (prasadtalasila) [inline cli/dtaas.toml:15]: user1,user2,user3

File: cli/src/cmd.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+@user.command()
+def delete():
+    """
+    removes the USERNAME user from DTaaS\n
+    Specify the users in dtaas.toml [users].delete\n
+    """
+
+    configObj = dtaasPkg.Config()
+    
+    err = userPkg.deleteUser(configObj)
```
Comment (prasadtalasila) [inline cli/src/cmd.py:49]: Please use try...catch structure

File: cli/src/pkg/dtaas.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
+import click
+from . import utils
+
+class Config:
+    def __init__(self):        
+        config, err = utils.importToml('dtaas.toml')
+        if err!=None:
```
Comment (prasadtalasila) [inline cli/src/pkg/dtaas.py:7]: Is try....catch iseful here?

File: cli/src/pkg/users.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
+import subprocess
+from . import utils
+
+def __getComposeConfig(username, server, DTaaSDirPath):
```
Comment (prasadtalasila) [inline cli/src/pkg/users.py:4]: Nice notation for internal functions.

**Review by prasadtalasila — COMMENTED** [review]
I need to check the working of the code. Please do add the developer (poetry...) commands to the README

File: cli/src/pkg/users.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        return err
+    userList, err = configObj.getAddUsersList()
+    if err is not None:
+        return err
+    server, err= configObj.getServerDNS()
+    if err is not None:
+        return err
+    path, err = configObj.getPath()
+    if err is not None:
+        return err
```
Comment (astitva1905) [inline cli/src/pkg/users.py:75]: too many returns @prasadtalasila

**Comment by prasadtalasila** [comment]
> Code Climate has analyzed commit [0516468](https://github.com/INTO-CPS-Association/DTaaS/commit/0516468acba7374e160f78696f999dee689f4eea) and detected **24 issues** on this pull request.
> 
> Here's the issue category breakdown:
> Category 	Count
> Complexity 	12
> Duplication 	8
> Style 	4
> 
> View more on [Code Climate](https://codeclimate.com/github/INTO-CPS-Association/DTaaS/pull/578).

@astitva1905 please resolve these codeclimate issues as well. Thanks.

**Comment by prasadtalasila** [comment]
@astitva1905 some feedback on the CLI after trying in `localhost` scenario.

1. Can short option flags be added?
    ```
    short flags:
      -h or --help
    ```
1. Handle the condition where the DTaaS is not installed
1. `docker/.env` has three users while compose has only two users. Can it be deleted?
1. what happens when new user is added without restarting **forward-auth** container?
1. DTaaS CLI does not add user directories in `files/`. This needs to be done in CLI itself? If the directory does exist, the directory structure needs to be checked.
1. When a user is added, add array emptied and add user is started again, the console shows the recent user being added.
1. When delete user is executed, even with empty user array being empty, the user added recently is deleted
1. When a user from original `docker-compose` file is attempted to be deleted, the following error comes:
    ```
    dtaas admin user delete
    no such service: kelvin
    Error: Error while deleting users: failed to stop containers
    ```

File: cli/DEVELOPER.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+To perform linting checks on the code:
+
+```bash
+pylint cli/src
+```
+
+## Testing
+
+Write automation tests for the new code written in the
```
Comment (prasadtalasila) [inline cli/DEVELOPER.md:96]: a separate `test.toml` file

File: cli/DEVELOPER.md [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+_cli/tests_ directory. Be sure to name any new files as
+_test_*_.py_. To run all tests, with coverage:
+
+```bash
```
Comment (prasadtalasila) [inline cli/DEVELOPER.md:100]: concise format
https://github.com/INTO-CPS-Association/DTaaS/blob/feature/distributed-demo/servers/execution/runner/DEVELOPER.md#hammer_and_wrench-developer-commands

**Pull request merged**